iml-services/iml-corosync/src/main.rs

```diff
@@ -0,0 +1,248 @@
+use iml_tracing::tracing;
+use iml_wire_types::{
+    high_availability::{Cluster, Node},
+    Fqdn,
+};
+use thiserror::Error;
+
+// Default pool limit if not overridden by POOL_LIMIT
+const DEFAULT_POOL_LIMIT: u32 = 2;
+
+#[tokio::main]
+async fn main() -> Result<(), ImlCorosyncError> {
+    iml_tracing::init();
+
+    let rabbit_pool = iml_rabbit::connect_to_rabbit(1);
+
+    let conn = iml_rabbit::get_conn(rabbit_pool).await?;
+
+    let ch = iml_rabbit::create_channel(&conn).await?;
+
+    let mut s = consume_data::<Cluster>(&ch, "rust_agent_corosync_rx");
+
+    let pool = get_db_pool(get_pool_limit().unwrap_or(DEFAULT_POOL_LIMIT)).await?;
+
+    sqlx::migrate!("../../migrations").run(&pool).await?;
+
+    while let Some((fqdn, cluster)) = s.try_next().await? {
+        let host_id = get_host_id_by_fqdn(&fqdn, &pool).await?;
+
+        let host_id = match host_id {
+            Some(id) => id,
+            None => {
+                tracing::warn!("Host '{}' is unknown, discarding incoming data", fqdn);
+
+                continue;
+            }
+        };
+
+        let (node_ids, node_names): (Vec<String>, Vec<String>) = cluster
+            .nodes
+            .iter()
+            .map(|x| (x.id.to_string(), x.name.to_string()))
+            .unzip();
+
+        // Delete old rows
+        sqlx::query!(
+            r#"
+            DELETE FROM corosync_node
+            USING corosync_node_managed_host
+            WHERE id = corosync_node_id AND name = corosync_node_name
+            AND corosync_node_id != ALL($1)
+            AND corosync_node_name != ALL($2)
+            AND host_id = $3
+        "#,
+            &node_ids,
+            &node_names,
+            host_id
+        )
+        .execute(&pool)
+        .await?;
+
+        // Upsert the rest
+        upsert_cluster_nodes(cluster.nodes, &pool).await?;
+
+        sqlx::query!(
+            r#"
+            INSERT INTO corosync_node_managed_host (host_id, corosync_node_id, corosync_node_name)
+            SELECT $1, corosync_node_id, corosync_node_name FROM UNNEST($2::text[], $3::text[])
+            AS t(corosync_node_id, corosync_node_name)
+            ON CONFLICT (host_id, corosync_node_id, corosync_node_name)
+            DO NOTHING
+            "#,
+            host_id,
+            &node_ids,
+            &node_names
+        )
+        .execute(&pool)
+        .await?;
+    }
+
+    Ok(())
+}
+
+#[derive(Error, Debug)]
+pub enum ImlCorosyncError {
+    #[error(transparent)]
+    ImlRabbitError(#[from] iml_rabbit::ImlRabbitError),
+    #[error(transparent)]
+    ImlServiceQueueError(#[from] ImlServiceQueueError),
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+    #[error(transparent)]
+    SqlxCoreError(#[from] sqlx::Error),
+    #[error(transparent)]
+    SqlxMigrateError(#[from] sqlx::migrate::MigrateError),
+}
+
+async fn get_host_id_by_fqdn(fqdn: &Fqdn, pool: &PgPool) -> Result<Option<i32>, ImlCorosyncError> {
```

Comment:
Might be nice to have these function ins a lib.rs file?

Reply:
could be but, I don't see a need to.